### ISSUE TITLE
New classes - campus, zone & device?

### ISSUE THREAD

**Issue by ghost** [Cetis/intelligent-campus#1]
The current dashboard uses the following data model:

Institution -> Campus -> Building -> Room -> Sensor -> Measure

I think we can map that well to the data model here with a few changes - so we end up with:

Institution (from LA) -> [Campus] -> Site -> Room / Location -> [Zone] -> [Device]

with the other bits like Feature hanging off Room too. Those in [] are not in the model yet and I'm wondering if you felt it worth adding them.

Campus is there as we thought that institutions may want to group sensor readings by campus and compare them - e.g. if the rooms are consistently hotter on Campus X but the heating bills are consistently lower can they learn something about the insulation on Campus X to improve Campus Y? Or in generaly student satisfaction is greater on Campus A - is there any IoT data that could help us understand why?

Zone is a grouping of sensors that may have a 1 to 1 relationship with a room, but in big open plan offices, large lecture halls and the like, users may want to group sensors into multiple zones.

A Device would be a grouping of sensors - so if you take an example from a device in the Jisc office - we have a single device that contains multiple sensors. Each sensor creates a measurement - temp, humidity, etc. We think that users may want to get reports on a device and that the device is the unit that will be replaced/need new batteries/need locating.

**Comment by ghost** [comment]
Just noticed - Campus would just be another type of Site. That makes sense - I'll go with that.

**Comment by scottbw** [comment]
I was going to say that but you beat me to it :D 

Vendors such as CELCAT tend to use "Site" as a generic term for all kinds of locations that can have rooms, so campuses, groups of buildings that form a distinct precinct in the campus, buildings, logical areas within a building etc. Which is why Site can be a hierarchy of Sites.

If a Zone spans multiple rooms, then it may be better represented as a Site?

**Comment by ghost** [comment]
:-) Cool - bit more flexible with types of Site I think so that is good.

Interesting point about zones. I don't think they will span multiple rooms - a room could span multiple zones however. e.g. in our Bristol office we have a big open plan area that could be described as a single room. However, the East side of that big room has some sensors and the West side too - I don't think it it is useful to fuse the readings for the sensors across the whole room - and it would confuse people who might start looking for walls or mess with the room map if we broke that into smaller rooms. So probably room -> zone but not the other way around.

**Comment by scottbw** [comment]
Sounds sensible - we should add Zone to the model.

I can think of a few spaces, like the ground floor of the business school at MMU, that don't easily translate into "rooms" but would make sense as zones - that particular space combines a lunchtime seating area, cafe area, casual seating area, reception area, and a large space near the doors often used for promotions and outreach. I can imagine wanting some idea of utilisation of each zone within the 'room' rather than for the space as a whole

**Comment by scottbw** [comment]
The only downside is that, while we tend to have good usable data from timetable systems about rooms and other bookable spaces, the only information on other more free-use spaces tends to come from estates management tools like Archibus and Planone and tends to be more complex to interpret.

**Comment by ghost** [comment]
There will be complexity interpreting any incoming data I suspect - this supports the dashboard edit functions - we can import what we can and then provide UI for the institution to edit the rooms, add zones, etc.

**Comment by scottbw** [comment]
I'm thinking an 80/20 principle here - we pull in what is easiest to pull in (probably bookable rooms and sites from CELCAT or Scientia) as a first stage, then look at other locations and zones later.

**Comment by scottbw** [comment]
I've added Zone to the data model.

Now, what about Device? Isn't that part of the event data? How might it be used for analysis purposes as a UDD dimension?

**Comment by ghost** [comment]
If we can get room data from CELCAT or Scientia that would be a good start though I fear that has proved difficult in the past - mostly policy rather than technical issues.

Device is the link between a zone and the event data and can carry additional data about the device that may be important - such as the type of device used (a PIR would record motion differently to an ultrasonic sensor) - that will not be part of the event itself.

The sensors we have seen so far do not report anything about their hardware other than some metadata on the communication used and perhaps battery power level and so we need somewhere to put that information.

An alternative would be a list of deviceEUIs in a Zone but that feels messy and limits what we can record about the Device itself.

Conceptually there is a difference between a Device and a Sensor - a Device may consist of multiple sensors - as is the case with Safehouse 1 & 2. The sensors fire the event and we record a sensor event (the temperature) rather than the device event (temperature, humidity, light level, noise level, CO2).

**Comment by scottbw** [comment]
I think in this case CELCAT/Scientia data is critical. Its the only system that links learner activity and the physical estate.

I'm not sure about devices - it seems more like a BMS case than an IC case. I'm struggling to come up with a good reason we'd want a provider to give us a list of devices, rather than just the measurement events related to building spaces. Adding devices and sensors feels like we're modelling a BMS, rather than the dimensions needed for analytics.

**Comment by ghost** [comment]
Agree re CELCAT/Scientia data - I have heard requests for it from other parts of Jisc too - we'll see.

"I'm not sure about devices - it seems more like a BMS case than an IC case." - you make a good point. I think I probably am coming at it from the BMS case and happy to not have it.

This still leaves the question how do we associate a sensor event/events with a zone though? It looks like this has to happen in the event rather than the zone, but the sensor will know little of where it is or the UDD its measures relate to? If that is true, we'll need to augment the sensor event on arrival with additional information - making the ingest of sensor data more complex - lookups per message, etc.

**Comment by alanepaull** [comment]
In the Sensors spec at the moment, we've have:

**Detector:** A single part of a Measurement System defining sampling and response characteristics of a detection device. A Detector has only 1 input and 1 output, both being scalar quantities. A Detector senses 1 type of measurement. A Sensor may contain several Detectors. Data is transmitted at Sensor level, a Detector is not separately identified.

**Sensor:** An entity capable of observing phenomena and returning observed values. A Sensor may contain several Detectors.

So, what you've called Devices above, we've called Sensors. And what you've called Sensors above, we've called Detectors. We've been using the OGC SensorML standard.

In the details of the measurement_event spec, the measurement entity covers the single detected instance (for example, temperature), while the sensor entity covers the equipment (so this entity contains the device details, such as MAC address / IP address, ID for the device, and so on.

The JSON with only mandatory elements (our minimum data proposal) has this:
```
{
"measurement_event": {
    "context": {
        "id": "kccBCAICAQYR/zMBF2QOEAABARoB1gCfAQAJCUQ2QjI3ODiP",
        "location": {
            "room_id": "AS34"
        },
        "timestamp": "2018-07-23T16:27:10.418789Z"
    },
    "measurement": {
        "id": 129,
        "value": 27.5
	},
	"sensor": {
		"id": "15"
	}
}
}
```

How does this fit with practicalities?
